### PR TITLE
Add Database support to Login Server

### PR DESCRIPTION
**Pull request by kamilprzyb**
Login Server now uses a database to save accounts and characters

1. Client has to be modified so it doesn't skip loging screen
2. Login has to be 5-20 characters, it can only contain letters, numbers, dash and underscore
3. Password has to be 8-20 characters, it can contain letters numbers and special characters
4. If you get "Login error" message, it probably means you're failing points 2 and 3
5. When you enter a Login for the first time, new account will be created. This will be used as a registration system for now
6. Your character is only saved during character creation, and after finishing tutorial (for now)

### CODE REVIEW

File: src/Database.cpp [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+	sqlite3pp::query::iterator i = qry.begin();
+	const char* actual;
+	std::tie(actual) =
+		(*i).get_columns<const char*>(0);
+	return BCrypt::validatePassword(password, actual);
+		
+}
+
+int Database::getUserID(std::string login) {
+	std::string q = "SELECT AccountID FROM Accounts WHERE Login = \"" + login + "\"";
```
Comment (Raymonf) [inline src/Database.cpp:101]: SQLi

File: src/Database.cpp [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+
+	sqlite3pp::query::iterator i = qry.begin();
+	int result;
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+	return result;
+}
+
+int Database::getUserSlotsNum(int AccountId) {
+	std::string q = "SELECT COUNT(PlayerID) FROM Players WHERE AccountID = "+ std::to_string(AccountId);
```
Comment (Raymonf) [inline src/Database.cpp:115]: This is "theoretically fine" but should very much be using prepared statements as well.

File: src/Database.cpp [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+	sqlite3pp::query::iterator i = qry.begin();
+	int result;
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+	return result;
+}
+
+bool Database::isNameFree(std::string First, std::string Second) {
+	std::string q = "SELECT COUNT(PlayerID) FROM Players WHERE FirstName = \""
```
Comment (Raymonf) [inline src/Database.cpp:129]: Also here.

File: src/Database.cpp [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+	return (result == 0);
+}
+
+void Database::createCharacter(sP_CL2LS_REQ_SAVE_CHAR_NAME* save, int AccountID) {
+	std::string charData = CharacterToJson(save);
+
+	sqlite3pp::command cmd(
+		db, "INSERT INTO Players (AccountID, Slot, FirstName, LastName, CharData) VALUES (:AccountID, :Slot, :FirstName, :LastName, :CharData)"
```
Comment (Raymonf) [inline src/Database.cpp:147]: This is good.

File: src/Database.cpp [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+	cmd.bind(":id", id);
+	cmd.execute();
+
+}
+
+void Database::finishTutorial(int PlayerID) {
+
+	std::string json;
+
+	std::string q = "SELECT CharData FROM Players WHERE PlayerID = " + std::to_string(PlayerID);
```
Comment (Raymonf) [inline src/Database.cpp:181]: "theoretically fine"

File: src/Database.cpp [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+		
+	);
+
+	cmd.bind(":data", json, sqlite3pp::nocopy);
+	cmd.bind(":id", PlayerID);
+	cmd.execute();
+}
+
+int Database::getCharacterID(int AccountID, int slot) {
+	std::string q = "SELECT PlayerID FROM Players WHERE AccountID = " +
```
Comment (Raymonf) [inline src/Database.cpp:209]: "theoretically fine"

File: src/Database.cpp [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+	int result;
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+	return result;
+}
+
+int Database::deleteCharacter(int characterID, int accountID) {
+
+	//checking if requested player exist and is bound to the account
+	std::string q = "SELECT COUNT(AccountID) FROM Players WHERE AccountID = " + std::to_string(accountID)
```
Comment (Raymonf) [inline src/Database.cpp:226]: "theoretically fine"

File: src/Database.cpp [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+	);
+
+	sqlite3pp::query::iterator i = qry.begin();
+	int result;
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+
+	if (result>0) {
+		//get player character slot
+		q = "SELECT Slot FROM Players WHERE PlayerID = " + std::to_string(characterID);
```
Comment (Raymonf) [inline src/Database.cpp:240]: This is "theoretically fine" too

File: src/contrib/INIReader.hpp [hunk 9]
```diff
@@ -289,7 +289,7 @@ inline int ini_parse(const char* filename, ini_handler handler, void* user)
     FILE* file;
     int error;
 
-    file = fopen(filename, "r");
+    errno_t errorno = fopen_s(&file, filename, "r");
```
Comment (Raymonf) [inline src/contrib/INIReader.hpp:292]: Are these changes from a fork?

File: src/main.cpp [hunk 10]
```diff
@@ -29,7 +30,7 @@ int main() {
     }
 #endif
     settings::init();
-    std::cout << "[INFO] Protocol version: " << PROTOCOL_VERSION << std::endl;
+    //std::cout << "[INFO] Protocol version: " << PROTOCOL_VERSION << std::endl;
```
Comment (Raymonf) [inline src/main.cpp:33]: Why is this being removed?

File: src/structs/0728.hpp [hunk 11]
```diff
@@ -4312,574 +4312,574 @@ struct sP_LS2CL_REP_SHARD_LIST_INFO_SUCC {
...
-static_assert(sizeof(sPC_BATTERYs) == 12 || sizeof(sPC_BATTERYs) == 0);
-static_assert(sizeof(sPC_NanoSlots) == 16 || sizeof(sPC_NanoSlots) == 0);
-static_assert(sizeof(sPC_Nano) == 12 || sizeof(sPC_Nano) == 0);
-static_assert(sizeof(sNPCLocationData) == 24 || sizeof(sNPCLocationData) == 0);
-static_assert(sizeof(sGroupNPCLocationData) == 44 || sizeof(sGroupNPCLocationData) == 0);
-static_assert(sizeof(sPCGroupMemberInfo) == 112 || sizeof(sPCGroupMemberInfo) == 0);
-static_assert(sizeof(sNPCGroupMemberInfo) == 32 || sizeof(sNPCGroupMemberInfo) == 0);
-static_assert(sizeof(sEPElement) == 36 || sizeof(sEPElement) == 0);
-static_assert(sizeof(sChannelInfo) == 8 || sizeof(sChannelInfo) == 0);
+static_assert(sizeof(sP_CL2FE_REQ_PC_ENTER) == 80 || sizeof(sP_CL2FE_REQ_PC_ENTER) == 0, "");
```
Comment (Raymonf) [inline src/structs/0728.hpp:4315]: ?

Any modern C++17-supporting compiler should do fine with what was there?

File: .gitignore [hunk 12]
```diff
@@ -1,12 +1,16 @@
 .vscode/
+.vs/*
```
Comment (Raymonf) [inline .gitignore:2]: Is this from a fork? This shouldn't be pushed upstream.

File: OpenFusion.sln [hunk 13]
```diff
@@ -0,0 +1,31 @@
+﻿
```
Comment (Raymonf) [inline OpenFusion.sln:1]: Is this from a fork? This shouldn't be pushed upstream. [2]

File: OpenFusion.vcxproj [hunk 14]
```diff
@@ -0,0 +1,207 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Raymonf) [inline OpenFusion.vcxproj:1]: Is this from a fork? This shouldn't be pushed upstream. [3]

File: OpenFusion.vcxproj.filters [hunk 15]
```diff
@@ -0,0 +1,204 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Raymonf) [inline OpenFusion.vcxproj.filters:1]: Is this from a fork? This shouldn't be pushed upstream. [4]

File: OpenFusion.vcxproj.user [hunk 16]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Raymonf) [inline OpenFusion.vcxproj.user:1]: Is this from a fork? This shouldn't be pushed upstream. [5]

File: src/CNLoginServer.hpp [hunk 17]
```diff
@@ -9,6 +9,19 @@
 struct CNLoginData {
     std::map<int64_t, Player> characters;
     int64_t selectedChar;
+    int UserID;
+};
+
+enum LOGINERRORID {
```
Comment (Raymonf) [inline src/CNLoginServer.hpp:15]: enum class

File: src/CNProtocol.cpp [hunk 18]
```diff
@@ -298,7 +298,7 @@ void CNServer::start() {
                 continue;
             }
 
-            std::cout << "New connection! " << inet_ntoa(address.sin_addr) << std::endl;
+            //std::cout << "New connection! " << inet_ntoa(address.sin_addr) << std::endl;
```
Comment (Raymonf) [inline src/CNProtocol.cpp:301]: ?

File: src/Database.cpp [hunk 19]
```diff
@@ -0,0 +1,460 @@
...
+		file.close();
+		//if exists, assign it		
+		db = sqlite3pp::database ("data.db");
+
+		DEBUGLOG(
+			std::cout << "Database in operation" << std::endl;
+		)
+	}
+	else {
+		//if doesn't, create all the tables
```
Comment (Raymonf) [inline src/Database.cpp:37]: This is fine for now, but we should probably have a separate table called "DatabaseVersion" or something that keeps track. Something like a primitive database migration system.

File: src/Database.cpp [hunk 20]
```diff
@@ -0,0 +1,460 @@
...
+
+	cmd.bind(":login", login, sqlite3pp::nocopy);
+	cmd.bind(":password", password, sqlite3pp::nocopy);
+	cmd.execute();
+	
+}
+
+bool Database::doesUserExist(std::string login) {
+
+	std::string q = "SELECT COUNT(AccountID) FROM Accounts WHERE Login = \"" + login + "\"";
```
Comment (Raymonf) [inline src/Database.cpp:72]: SQLi

File: src/Database.cpp [hunk 21]
```diff
@@ -0,0 +1,460 @@
...
+	
+	sqlite3pp::query::iterator i = qry.begin();
+	int result;
+	std::tie(result) =
+		(*i).get_columns<int>(0);
+	return (result != 0);
+}
+
+bool Database::isPasswordCorrect(std::string login, std::string password) {
+	std::string q = "SELECT Password FROM Accounts WHERE Login = \"" + login + "\"";
```
Comment (Raymonf) [inline src/Database.cpp:86]: SQLi

File: src/Database.cpp [hunk 22]
```diff
@@ -0,0 +1,460 @@
...
+		DEBUGLOG(
+			std::cout << "Done" << std::endl;
+		)
+	}
+	
+
+}
+
+bool Database::isLoginDataGood(std::string login, std::string password) {
+	std::regex loginRegex("^([A-Za-z\\d_\\-]){5,20}$");
```
Comment (Raymonf) [inline src/Database.cpp:53]: I reckon the regex check here is to "prevent" SQLi, but we can't guarantee that this code isn't touched by anyone else. If this check is somehow changed or removed, the entire database would be in shambles.

File: src/contrib/bcrypt/bcrypt.c [hunk 23]
```diff
@@ -0,0 +1,230 @@
+/*
```
Comment (Raymonf) [inline src/contrib/bcrypt/bcrypt.c:1]: I don't think we need this file

**Comment by kamilprzyb** [comment]
> Did you base your work off of a fork? We shouldn't have any VS solution files here.
I based it off Joseph's fork

File: src/contrib/INIReader.hpp (hunk 9 above)
Comment (kamilprzyb) [reply src/contrib/INIReader.hpp:292]: yes

File: src/main.cpp (hunk 10 above)
Comment (kamilprzyb) [reply src/main.cpp:33]: fork

File: src/structs/0728.hpp (hunk 11 above)
Comment (kamilprzyb) [reply src/structs/0728.hpp:4315]: I haven't done that

File: .gitignore (hunk 12 above)
Comment (kamilprzyb) [reply .gitignore:2]: fork

File: OpenFusion.sln (hunk 13 above)
Comment (kamilprzyb) [reply OpenFusion.sln:1]: fork

File: OpenFusion.vcxproj (hunk 14 above)
Comment (kamilprzyb) [reply OpenFusion.vcxproj:1]: fork

File: OpenFusion.vcxproj.filters (hunk 15 above)
Comment (kamilprzyb) [reply OpenFusion.vcxproj.filters:1]: fork

File: OpenFusion.vcxproj.user (hunk 16 above)
Comment (kamilprzyb) [reply OpenFusion.vcxproj.user:1]: fork

File: src/CNLoginServer.hpp (hunk 17 above)
Comment (kamilprzyb) [reply src/CNLoginServer.hpp:15]: gonna fix it

File: src/CNProtocol.cpp (hunk 18 above)
Comment (kamilprzyb) [reply src/CNProtocol.cpp:301]: fork

**Comment by kamilprzyb** [comment]
I'm going to rewrite db connection using some ORM library later, so there is no worry about injections and it's a lot cleaner to work upon. It should also do migrations.

File: src/Database.cpp (hunk 19 above)
Comment (FinnHornhoover) [reply src/Database.cpp:37]: +1 I second the DatabaseVersion table idea.

**Comment by CPunch** [comment]
closed duplicate

**Pull request closed without merging**